### PR TITLE
feat: tests for csm build report

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

File: src/modules/csm/csm.py [hunk 1]
```diff
@@ -110,18 +108,24 @@ def build_report(self, blockstamp: ReferenceBlockStamp) -> tuple:
         else:
             logger.info({"msg": "No previous CID available"})
 
-        tree = self.make_tree(shares)
-        tree_cid: CID | None = None
+        if distributed > 0:
+            curr_tree = self.make_tree(shares)
+            if not curr_tree:
+                raise InconsistentData("No tree to publish but shares are distributed")
```
Comment (madlabman) [inline src/modules/csm/csm.py:114]: I'm not sure if it makes sense. `make_tree` doesn't depend on `distributed`. I think it's better to check on line 98 invariant `distributed and shares`.

File: src/modules/csm/types.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ class ReportData:
     consensusVersion: int
     ref_slot: SlotNumber
     tree_root: HexBytes
-    tree_cid: CID | Literal[""]
+    tree_cid: CID
```
Comment (madlabman) [inline src/modules/csm/types.py:22]: `prev_cid` still can be `Literal[""]` and hence zeroed first report will result in `tree_cid = ""`.

File: src/web3py/extensions/csm.py [hunk 3]
```diff
@@ -46,10 +46,8 @@ def get_csm_last_processing_ref_slot(self, blockstamp: BlockStamp) -> SlotNumber
     def get_csm_tree_root(self, blockstamp: BlockStamp) -> HexBytes:
         return self.fee_distributor.tree_root(blockstamp.block_hash)
 
-    def get_csm_tree_cid(self, blockstamp: BlockStamp) -> CID | None:
+    def get_csm_tree_cid(self, blockstamp: BlockStamp) -> CID:
         result = self.fee_distributor.tree_cid(blockstamp.block_hash)
-        if not result:
-            return None
```
Comment (madlabman) [inline src/web3py/extensions/csm.py:52]: Why so? Empty string is not a valid CID and will result in CIDv1("") which is not correct.

File: tests/conftest.py [hunk 4]
```diff
@@ -113,6 +114,9 @@ def keys_api_client(request, responses_path, web3):
 @pytest.fixture()
 def csm(web3):
     mock = Mock()
+    mock.module.MAX_OPERATORS_COUNT = UINT64_MAX
+    web3.ipfs = Mock()
+    web3.lido_contracts = Mock()
```
Comment (madlabman) [inline tests/conftest.py:119]: Why not to keep it at the test level?

File: tests/modules/csm/test_csm_module.py [hunk 5]
```diff
@@ -493,3 +498,249 @@ def test_collect_data_fulfilled_state(
...
+
+@dataclass(frozen=True)
+class BuildReportTestParam:
+    prev_root: str
+    prev_cid: str
+    prev_acc_shares: Iterable[tuple[NodeOperatorId, int]]
+    curr_distribution: Mock
+    curr_root: str
+    curr_cid: str
+    curr_log: str
```
Comment (madlabman) [inline tests/modules/csm/test_csm_module.py:511]: We have two CIDs, tree and log. It's not very clear from this params.

File: tests/modules/csm/test_csm_module.py [hunk 6]
```diff
@@ -493,3 +498,249 @@ def test_collect_data_fulfilled_state(
...
+                        # distributed
+                        6,
+                        # shares
+                        defaultdict(int, {NodeOperatorId(0): 1, NodeOperatorId(1): 2, NodeOperatorId(2): 3}),
+                        # log
+                        Mock(),
+                    )
+                ),
+                curr_root="0x100e",
+                curr_cid="0x100c",
```
Comment (madlabman) [inline tests/modules/csm/test_csm_module.py:558]: It looks like bytes instead of CID.

File: src/modules/csm/csm.py (hunk 1 above)
Comment (madlabman) [reply src/modules/csm/csm.py:114]: It's guaranteed by `calculate_distribution` by the way.

File: tests/modules/csm/test_csm_module.py (hunk 6 above)
Comment (vgorkavenko) [reply tests/modules/csm/test_csm_module.py]: fixed

File: tests/modules/csm/test_csm_module.py (hunk 5 above)
Comment (vgorkavenko) [reply tests/modules/csm/test_csm_module.py]: fixed

File: tests/conftest.py (hunk 4 above)
Comment (vgorkavenko) [reply tests/conftest.py]: fixed

File: src/web3py/extensions/csm.py (hunk 3 above)
Comment (vgorkavenko) [reply src/web3py/extensions/csm.py:52]: changed

File: src/modules/csm/csm.py (hunk 1 above)
Comment (vgorkavenko) [reply src/modules/csm/csm.py]: changed

File: src/modules/csm/csm.py [hunk 7]
```diff
@@ -98,7 +99,7 @@ def build_report(self, blockstamp: ReferenceBlockStamp) -> tuple:
         prev_root = self.w3.csm.get_csm_tree_root(blockstamp)
         prev_cid = self.w3.csm.get_csm_tree_cid(blockstamp)
 
-        if bool(prev_root) != bool(prev_cid):
+        if bool(prev_root) != (prev_root is not ZERO_HASH):
```
Comment (madlabman) [inline src/modules/csm/csm.py:102]:
```suggestion
        if bool(prev_cid) != (prev_root is not ZERO_HASH):
```

File: src/modules/csm/csm.py [hunk 8]
```diff
@@ -110,13 +111,13 @@ def build_report(self, blockstamp: ReferenceBlockStamp) -> tuple:
...
                 blockstamp.ref_slot,
-                tree_root=prev_root or HexBytes(32),
+                tree_root=prev_root,
                 tree_cid=prev_cid or "",
                 log_cid=log_cid,
                 distributed=0,
             ).as_tuple()
 
-        if prev_cid and prev_root:
+        if prev_cid and prev_root is not ZERO_HASH:
```
Comment (madlabman) [inline src/modules/csm/csm.py:120]: Didn't quite get why the check for root at all.

File: src/modules/csm/csm.py (hunk 7 above)
Comment (madlabman) [reply src/modules/csm/csm.py:102]: If I'm right, there's a missing test case.

File: src/modules/csm/csm.py (hunk 8 above)
Comment (vgorkavenko) [reply src/modules/csm/csm.py:120]: practically we expect them both to be presented or not presented so i've tried to show it here

**Review by madlabman — APPROVED** [review]

**Pull request merged**